### ISSUE TITLE
Σ₀ continual-training loop — close the flywheel (harvest→verify→train→eval→promote)

### ISSUE THREAD

**Issue by alex-place** [alex-place/lantern-os#911]
## Σ₀ continual-training loop — close the flywheel

Turn the Σ₀ Ouro Coder's hand-cranked train→promote into a **closed, gated loop**: the
system's own execution-verified coding successes become training data → a new QLoRA adapter
is trained + evaluated → promoted to the live `final/` adapter **only on a measured HumanEval
`pass@1` win**.

**Handoff:** `docs/handoffs/sigma0-continual-training-2026-06-21.md`
**Design doc (KC):** `docs/SIGMA0-CONTINUAL-TRAINING.md`

### Status: ✅ data + decision half built, reviewed, tested · ⏳ GPU loop + live emitter outstanding

**Built (CPU stages green):**
- `scripts/harvest_coding_corpus.py` — gathers `{fn,instruction,code,asserts}` candidates; AST-based normalization, name-collision disambiguation, builtin-shadow guard.
- `scripts/continual_ouro_pipeline.py` — harvest → **execution-verify** → train → eval → **eval-gated promote** (crash-safe swap + append-only Convergence-Record ledger; pure unit-tested `decide_promotion`).
- `docs/SIGMA0-CONTINUAL-TRAINING.md` + a Knowledge Center link card (render-verified).

**Two ground-truth gates:** (1) only a green subprocess (`returncode==0`) enters training; (2) a new adapter replaces `final/` only if `pass@1` beats the incumbent by `--margin`.

**Test evidence:** `--self-test` 6/6 (incl. bootstrap); harvest 192/192 (6 disambiguated), execution-verify 191/192, base-dedup → 250 honest rows (was 428 w/ 184 dupes); crash-safe promote unit-tested incl. **forced mid-swap failure → `final/` restored intact**; 6 normalization edge cases pass.

**Adversarial review:** 4-dimension workflow, 15 confirmed findings → 14 fixed (2 HIGH loop-breakers: non-atomic promote swap; `<out>` vs `<out>/final` adapter path), 1 deferred (see #6 below).

### Outstanding work
- [ ] **1. Run the full GPU loop once** — `continual_ouro_pipeline.py --train --eval --promote --full` on the `.venv-train` GPU box (stop `ouro_serve.py` first). Confirm promote-on-win / reject-leaves-`final`-untouched and a Convergence Record in `data/eval/ouro-promotion-log.jsonl`.
- [ ] **2. Record the rerun number** — full-HumanEval `ouro-final-rerun-full` `pass@1` (eval was in-flight at ~73% / 49 of 164) into the "Measured baseline" section of the design doc.
- [ ] **3. Wire the live-run emitter** (the real "emergent" closure) — autowork/keystone/chat coding successes emit `{fn,instruction,code,asserts}` rows for `harvest_coding_corpus.py --source-jsonl`. ⚠️ Keep OFFLINE: emit/log only; never trigger a retrain from the live request loop.
- [ ] **4. Restart the cached Ouro chat server** after the eval (`OURO_PORT=11434 .venv-train/Scripts/python scripts/ouro_serve.py`) — stopped to free the GPU for the rerun.
- [ ] **5. (Optional) Schedule the offline loop** (Windows scheduled task / cron) — stays opt-in + offline.
- [ ] **6. (Deferred review finding) `_FORBIDDEN_TOKENS` substring blocklist** in `build_ouro_coding_dataset.py` can false-positive on identifiers like `parse_datetime`/`reopen`. Low/latent (0 triggers today). Make AST/word-boundary aware when the live source feeds richer code.
- [ ] **7. Decide artifact tracking** — `data/ouro-harvest-candidates.jsonl` + `models/lantern-sigma0-coder/training-data.harvested.jsonl` are generated + not gitignored; likely gitignore (reproducible/agent-derived).

### Boundary (non-negotiable)
QLoRA is weight modification; the North Star forbids the **live** loop from retraining. This
pipeline is offline + opt-in + outside the request path by design. Closing #3 must preserve
this: harvest/log live, retrain/promote offline.

**Comment by alex-place** [comment]
## Item #3 complete — live-run emitter landed (#1040)

 is now wired into both autowork dispatch routes ( and ). When autowork commits verified (tests-passed) code, it appends a row to  that the offline pipeline can consume via:

```
python scripts/harvest_coding_corpus.py --source-jsonl data/harvest/live-coding-successes.jsonl
```

File is gitignored (agent-derived, local only). Never blocks the live path, never triggers retraining.

**Outstanding from #911:**
- [ ] Run the full GPU loop once (needs `ouro_serve.py` + GPU)
- [ ] Record the rerun pass@1 baseline in the design doc
- [ ] Restart the cached Ouro chat server after eval
- [ ] (Optional) Schedule the offline loop
- [ ] `_FORBIDDEN_TOKENS` AST hardening (low/latent)

**Comment by alex-place** [comment]
**Status sweep — 2026-06-22**

| Item | Status |
|------|--------|
| #3 Wire live emitter | ✅ Already done — `lib/harvest-emitter.js` exists; `routes/convergence-dispatch.js:276,651` calls `emitCodingSuccess()` on verified coding successes. `data/harvest/live-coding-successes.jsonl` is gitignored. |
| #7 Gitignore generated files | ✅ Fixed now — added `models/*/training-data.harvested.jsonl` to `.gitignore` (pushed to master `d204bd0b`). `data/ouro-harvest-candidates.jsonl` and `data/harvest/live-coding-successes.jsonl` were already there. |
| #1 Run full GPU loop | ⬜ GPU task — requires stopping `ouro_serve.py`, running `continual_ouro_pipeline.py --train --eval --promote --full`. Can't automate. |
| #2 Record rerun number | ⬜ Depends on #1. |
| #4 Restart Ouro server | ⬜ After #1 completes. |
| #5 Schedule offline loop | ⬜ Optional / opt-in. |
| #6 `_FORBIDDEN_TOKENS` AST fix | ⬜ Deferred — 0 triggers today. |

Outstanding items are all GPU-execution tasks or explicitly deferred. Leaving open until #1/#2/#4 are completed.
